table.go

```diff
@@ -76,61 +76,32 @@ func (rt *RoutingTable) GenRandPeerID(bucketID int) (peer.ID, error) {
 	bucketLen := len(rt.Buckets)
 	rt.tabLock.RUnlock()
 
-	var targetCpl int
-	if bucketID >= bucketLen-1 {
-		targetCpl = bucketLen
+	var targetCpl uint
+	if bucketID > (bucketLen - 1) {
+		targetCpl = uint(bucketLen) - 1
 	} else {
-		targetCpl = bucketID
+		targetCpl = uint(bucketID)
 	}
 
-	// generate random 16 bits
-	r := rand.New(rand.NewSource(time.Now().UnixNano()))
-	buf := make([]byte, 2)
-	_, err := r.Read(buf)
-	if err != nil {
-		return "", err
+	// We can only handle 16 bit prefixes.
+	if targetCpl > 16 {
+		targetCpl = 16
 	}
 
-	// replace the first targetCPL bits with those from the hashed local peer ID & toggle the (targetCpl+1)th bit
-	// so that exactly targetCpl bits match
-	numBytes := targetCpl / 8 // number of bytes we need to replace
-	numBits := targetCpl % 8  // number of bits we need to replace after numBytes have been replaced
+	// Extract the local prefix and a random prefix.
+	localPrefix := binary.BigEndian.Uint16(rt.local)
+	randPrefix := uint16(rand.Uint32())
 
-	// replace the bytes
-	byteIndex := 0
-	for ; byteIndex < numBytes; byteIndex++ {
-		buf[byteIndex] = rt.local[byteIndex]
-	}
-
-	// replace the bits
-	if byteIndex < len(buf) {
-		dstByte := buf[byteIndex]
-		srcByte := rt.local[byteIndex]
-		j := uint(7)
-		for k := 1; k <= numBits; k++ {
-			if isSet(srcByte, j) {
-				dstByte = setBit(dstByte, j)
-			} else {
-				dstByte = clearBit(dstByte, j)
-			}
-			j--
-		}
-
-		// toggle the next bit
-		if isSet(srcByte, j) {
-			dstByte = clearBit(dstByte, j)
-		} else {
-			dstByte = setBit(dstByte, j)
-		}
-		buf[byteIndex] = dstByte
-	}
+	// Combine the local prefix and the random bits at the correct offset
+	// such that the first `bucketID` bits match the local ID.
+	mask := (^uint16(0)) << targetCpl
+	targetPrefix := (localPrefix & mask) | (randPrefix & ^mask)
 
```

Comment:
@Stebalien  Had to make two changes to get this working correctly. 

1. CPL should be EXACTLY targetCpl. To do that, we should toggle the (targetCpl + 1)<sup>th</sup> bit in localPrefix & then copy (targetCpl + 1) bits from localPrefix to randPrefix.
2. mask should be generated as ```mask := (^uint16(0)) << (16 - target)``` because we want to copy target bits from the localPrefix and NOT from the randPrefix.